### PR TITLE
ENH: Move PyVectorContainer functionality into ITK

### PR DESCRIPTION
**Pull request by Leengit**
The functionality from ITK module `ITKMeshToPolyData` is being moved into `ITK` proper.

## PR Checklist
- [x] No [API changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#breaking-changes) were made (or the changes have been approved)
- [x] No [major design changes](https://github.com/InsightSoftwareConsortium/ITK/blob/master/CONTRIBUTING.md#design-changes) were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [ ] Updated API documentation (or API not changed)
- [x] Added [license](https://github.com/InsightSoftwareConsortium/ITK/blob/master/Utilities/KWStyle/ITKHeader.h) to new files (if any)
- [x] Added Python wrapping to new files (if any) as described in [ITK Software Guide](https://itk.org/ItkSoftwareGuide.pdf) Section 9.5
- [ ] Added [ITK examples](https://github.com/InsightSoftwareConsortium/ITKExamples) for all new major features (if any)

**Help please:** I am getting ctest errors for the moved code that are cryptic to me and I would appreciate feedback.  The output from the 3 failing ctests are in the file: [ctest-V.txt](https://github.com/InsightSoftwareConsortium/ITK/files/5488435/ctest-V.txt)

### CODE REVIEW

**Comment by dzenanz** [comment]
```text
2659: Loading ITKPyBase... done
2659: Loading ITKCommon... done
2659: Loading ITKStatistics... done
2659: Loading ITKImageFilterBase... Template itk::CastImageFilter<itk::Image<unsignedlonglong,2>,itk::Image<signedchar,2>>
```
implies that `CastImageFilter` has been wrapped for the mentioned instantiations in one of the modules loaded earlier, probably ITKPyBase or ITKCommon.

```text
2664: WARNING: In /home/lee.newberg/git/ITK-pybuffer/Modules/Core/Common/src/itkProcessObject.cxx, line 852
2664: BlockMatchingImageFilter (0x8c97090): Input already "FeaturePoints" already required!
```
probably means that `BlockMatchingImageFilter::AddRequiredInputName("FeaturePoints")` is being called more than once.

**Review by dzenanz — COMMENTED** [review]
Is `itkMeshToPolyDataFilter.h` supposed to be a part of this PR? Or is that addition planned for a later commit?

**Comment by Leengit** [comment]
I apologize for not being specific.  I need help understanding what the following means, and some hints as to where/how one might fix it:
```
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerSC'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerSS'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerUC'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerULL'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerUS'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerF'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerD'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerUI'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerPF2'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerPF3'>: empty class
```

Also with the following, should I be wanting to wrap more for VTK or less for ITK?:
```
2664: Traceback (most recent call last):
2664:   File "PythonTypeTest.py", line 93, in <module>
2664:     t, w = create_and_test(i, create_method(i))
2664:   File "PythonTypeTest.py", line 46, in create_and_test
2664:     I = t.New()
2664:   File "~/git/ITK-pybuffer-build/Wrapping/Generators/Python/itk/itkVTKImageExportPython.py", line 1878, in New
2664:     obj = itkVTKImageExportIULL2.__New_orig__()
2664: RuntimeError: ~/git/ITK-pybuffer/Modules/Bridge/VTK/include/itkVTKImageExport.hxx:93:
2664: itk::ERROR: itk::ERROR: VTKImageExport(0xea829a0): Type currently not supported
```

Also, the following is completely unexpected for me.  If you think that fixes for the above two won't fix this one, any thoughts on what this is about?:
```
2767: Traceback (most recent call last):
2767:   File "WatershedSegmentation1Test.py", line 47, in <module>
2767:     writer.Update()
2767: RuntimeError: ~/git/ITK-pybuffer/Modules/Segmentation/ConnectedComponents/include/itkRelabelComponentImageFilter.hxx:194:
2767: itk::ERROR: itk::ERROR: RelabelComponentImageFilter(0x5a00aa0): Output voxel range exceeded for relabeling.  Too many objects of sufficient size found!
```

**Comment by Leengit** [comment]
> Is `itkMeshToPolyDataFilter.h` supposed to be a part of this PR? Or is that addition planned for a later commit?

In short, I don't know.  I thought it was appropriate, but I am discovering that I know a lot less about the needed work here than I previously understood.  :-(

I'll try to figure it out.

**Comment by dzenanz** [comment]
RelabelComponentImageFilter takes a binary image and gives each connected component a distinct label. The warning will be triggered if there is 256 or more objects, but `uchar` type is used for label numbers. Or 32768 or more object with `signed short` etc.

**Comment by dzenanz** [comment]
You probably want to wrap `itkVTKImageExport`, but I am not sure. I don't know about `empty class`.

File: Modules/Bridge/NumPy/wrapping/test/itkMeshToPolyDataFilterTest2.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+point2 = PointType()
+point2[0] = 1
+point2[1] = 1
+mesh.SetPoint(2, point2)
+point3 = PointType()
+point3[0] = 1
+point3[1] = 1
+mesh.SetPoint(3, point3)
+
+filt = itk.MeshToPolyDataFilter.New(mesh)
```
Comment (thewtex) [inline Modules/Bridge/NumPy/wrapping/test/itkMeshToPolyDataFilterTest2.py:43]: We can focus on the `PyVectorContainer` wrapping in this PR and leave out `MeshtoPolyData`.

**Review by thewtex — CHANGES_REQUESTED** [review]

**Comment by thewtex** [comment]
Missing the *Python.h* include directory:

https://open.cdash.org/viewBuildError.php?buildid=6859301

File: Modules/Bridge/NumPy/wrapping/test/itkMeshToPolyDataFilterTest2.py (hunk 1 above)
Comment (Leengit) [reply Modules/Bridge/NumPy/wrapping/test/itkMeshToPolyDataFilterTest2.py:43]: I will remove that file.

**Comment by Leengit** [comment]
> You probably want to wrap `itkVTKImageExport` ...

I will wrap it for `ULL`.

**Comment by Leengit** [comment]
> Missing the _Python.h_ include directory:
> 
> https://open.cdash.org/viewBuildError.php?buildid=6859301

I don't know why that file cannot find `Python.h` to include but other files in that directory can find `Python.h`.  Other files first `#undef _POSIX_C_SOURCE` and `#undef _XOPEN_SOURCE` so I am trying that, even though that doesn't seem important.  Any thoughts on why `Python.h` is not findable in only this one case / how to fix that?

**Comment by Leengit** [comment]
1.    I am not seeing the `Python.h` problem on my local build, but will see how the commit I just pushed does on GitHub.

 1.    I continue to get 
```
2659: 7 empty classes found
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerSS'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerUC'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerF'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerD'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerUI'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerPF2'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerPF3'>: empty class
``` 
from `ctest -V -R PythonFindEmptyClasses`.  I will continue to try to figure out what this means and how to fix it, but if you have any insights, please share! 

 1.    I continue to get
```
2664:   File "PythonTypeTest.py", line 93, in <module>
2664:     t, w = create_and_test(i, create_method(i))
2664:   File "PythonTypeTest.py", line 46, in create_and_test
2664:     I = t.New()
2664:   File "/home/lee.newberg/git/ITK-pybuffer-build/Wrapping/Generators/Python/itk/itkVTKImageExportPython.py", line 1122, in New
2664:     obj = itkVTKImageExportIULL2.__New_orig__()
2664: RuntimeError: /home/lee.newberg/git/ITK-pybuffer/Modules/Bridge/VTK/include/itkVTKImageExport.hxx:93:
2664: itk::ERROR: itk::ERROR: VTKImageExport(0x7689080): Type currently not supported
```
from `ctest -V -R PythonTypeTest` even though I have taken steps to wrap `VTKImageExportULL`.  Thoughts as to what is going here would be appreciated.

**Review by thewtex — CHANGES_REQUESTED** [review]
Todo: add test to CMake configuration

**Comment by Leengit** [comment]
The only outstanding ctest error I am getting is from `PythonFindEmptyClasses`.  It includes
```
2659: 7 empty classes found
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerSS'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerUC'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerF'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerD'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerUI'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerPF2'>: empty class
2659: <class 'itk.itkPyBufferPython.itkPyVectorContainerPF3'>: empty class
```

Any ideas what this is about?

**Comment by thewtex** [comment]
https://open.cdash.org/test/283906442

```
E
======================================================================
ERROR: test_NumPyBridge_VectorContainer (__main__.TestNumpyVectorContainerMemoryviewInterface)
Try to convert a itk.VectorContainer into a Numpy array and back.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/1/s/Modules/Bridge/NumPy/wrapping/test/itkPyVectorContainerTest.py", line 39, in test_NumPyBridge_VectorContainer
    arr = itk.PyVectorContainer[itk.F].array_view_from_vector_container(v1)
  File "/Users/runner/work/1/s-build/Wrapping/Generators/Python/itk/itkPyBufferPython.py", line 7078, in array_view_from_vector_container
    ndarrview  = numpy.asarray(memview).view(dtype = numpydatatype).reshape(shape).view(numpy.ndarray)
NameError: name 'numpy' is not defined

----------------------------------------------------------------------
Ran 1 test in 0.522s

FAILED (errors=1)
itkTestDriver: Process exited with return value: 1
```

Is there a missing `import numpy` statement? Ideally, we should change this to canonical `import numpy as np` and `np.asarray[...]`.

**Comment by Leengit** [comment]
> Is there a missing `import numpy` statement? Ideally, we should change this to canonical `import numpy as np` and `np.asarray[...]`.

The problem is that `import numpy as np` requires `np.` as a prefix rather than `numpy.`  I have just changed instances of the latter to be the former and force-pushed.

**Comment by Leengit** [comment]
The errors from the `Wrapping/Generators/Python/Tests/findEmptyClasses.py` test are addressed by adding `PyVectorContainer` to the list of classes to exclude.  This is justified because we do not expect any methods of `PyVectorContainer` to be in CamelCase (for C++); methods are instead in snake_case (for Python).  In this case the methods also start with an underscore, which is private / hidden by convention in Python.

Thank you: @thewtex, @dzenanz

**Review by thewtex — APPROVED** [review]
:tada:

**Comment by thewtex** [comment]
@Leengit please address the failing Windows Python test:

https://open.cdash.org/test/288470747

**Comment by Leengit** [comment]
> @Leengit please address the failing Windows Python test:
> 
> https://open.cdash.org/test/288470747

To be addressed by #2113.

**Pull request merged**